game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/suicide/RemoveUnits.java

```diff
@@ -0,0 +1,61 @@
+package games.strategy.triplea.delegate.battle.steps.change.suicide;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import lombok.AllArgsConstructor;
+import org.triplea.java.collections.CollectionUtils;
+
+@AllArgsConstructor
+public abstract class RemoveUnits implements BattleStep {
+  private static final long serialVersionUID = 1322821208196377684L;
+
+  private final BattleState battleState;
+
+  private final BattleActions battleActions;
+
+  protected void removeUnits(final IDelegateBridge bridge, final Predicate<Unit> unitMatch) {
+    final Collection<Unit> deadAttackers =
+        CollectionUtils.getMatches(
+            battleState.getUnits(BattleState.Side.OFFENSE),
+            unitMatch.and(Matches.unitIsSuicideOnAttack()));
+    final Collection<Unit> deadDefenders =
+        CollectionUtils.getMatches(
+            battleState.getUnits(BattleState.Side.DEFENSE),
+            unitMatch.and(Matches.unitIsSuicideOnDefense()));
+    bridge
+        .getDisplayChannelBroadcaster()
+        .deadUnitNotification(
+            battleState.getBattleId(),
+            battleState.getAttacker(),
+            deadAttackers,
+            getDependents(deadAttackers));
+    bridge
+        .getDisplayChannelBroadcaster()
+        .deadUnitNotification(
+            battleState.getBattleId(),
+            battleState.getDefender(),
+            deadDefenders,
+            getDependents(deadDefenders));
+    final List<Unit> deadUnits = new ArrayList<>(deadAttackers);
+    deadUnits.addAll(deadDefenders);
+    battleActions.remove(deadUnits, bridge, battleState.getBattleSite(), null);
```

Comment:
I think we should convert `battleActions` to be a callback instead. That would remove the side effects.  So instead of having a `BattleActions` we would have a `Consumer<Collection<Unit>>` that would be a callback to do the removal. This would make it a bit more clear from the caller how the battle actions parameter is mutated. Otherwise, it's a bit difficult to follow to realize that we're storing a reference and are modifying the caller by modifying a reference passed to us here. If that is the case, then the abstraction of this class is arguably not that helpful, you still have to understand this entire class and its side effects to understand the caller, which loses all information hiding.

Reply:
I want to refactor `remove` so that it doesn't have the side effects.  Most of the code in `remove` should be in the `BattleState` api since it is interacting with the `BattleState`.  The only part that should stay in the `BattleActions` api is just the communication with the bridge.